### PR TITLE
Add SAMS REST API client

### PR DESCRIPTION
**Pull request by chrsmith**
The SAMS SDK exposes a slick protobuf-based API that most SAMS integrators will use. However, if the SAMS client is only using SAMS for authentication, but not authorization, then a _different_ API is required.

The SAMS REST API only has one API endpoint, but it is (kinda surprisingly) the only way to make the "SAMS sans auth" work. When using the OAuth/OIDC portion of SAMS, the OAuth Token Introspection endpoint does not include enough information about the user who authorized the token to be useful. (i.e. their SAMS account ID and primary email address.)

The SAMS REST API's `/v1/user` endpoint however, does. Hence why it is being added here to the SAMS SDK.

This is needed for the Self-Serve Cody backend when it moves the UI into sourcegraph.com. The Self-Serve Cody backend will no longer be able to use the SAMS "session API", and instead will need to authorize requests on behalf of the end user, via their SAMS access token.

## Test plan

I tested the code locally, against my local SAMS server. Things worked as expected.

### CODE REVIEW

File: README.md
Comment (unknwon) [inline README.md:175]: Convention here is to use all lowercase letters 😁 
```suggestion
	samsrest "github.com/sourcegraph/sourcegraph-accounts-sdk-go/rest"
```

File: rest/client.go
Comment (unknwon) [inline rest/client.go:31]: I think we should ask a `oauth2.TokenSource` when creating the client for best compatibility (so the other subpackage `auth` can take advantage of this API in the idiomatic way), and in the README example, we could use `oauth2.StaticTokenSource(...)` as the simplest illustration setup.

File: rest/client.go
Comment (unknwon) [inline rest/client.go:27]: Could you unwrap this interface (and move docstrings to the implementation directly)? In our current practice, consumers should define their interface on-demand for mocking/interoperability.

File: rest/client.go
Comment (unknwon) [inline rest/client.go:45]: nit: In general I'd say `GetUser` is concise and clear enough (to get a `User` object), unless you think it otherwise.

```suggestion
func (c *client) GetUser(ctx context.Context, token string) (*User, error) {
```

Or `GetUserInfo`, which uses a consistent term (as it is hitting the user info endpoint).

File: rest/types.go
Comment (unknwon) [inline rest/types.go:1]: Do you think it's worth naming it `restv1` upfront? FWIW The Clients API SDK encodes the v1 directly.

**Review by unknwon — COMMENTED** [review]
🫡 Thanks for taking on the initiative!

File: README.md
Comment (chrsmith) [reply README.md:175]: Done

File: rest/client.go
Comment (chrsmith) [reply rest/client.go:27]: Done.

File: rest/client.go
Comment (chrsmith) [reply rest/client.go:45]: Sounds good to me. #namingishard I wasn't sure if you would have preferred "GetOIDCUserDetails" or something more specialized. But I agree `GetUser` is the most preferable option.

File: rest/types.go
Comment (chrsmith) [reply rest/types.go:1]: Good catch. I personally don't think it's worth naming it `restv1`, but it's probably better be consistent with all the other services/APIs that we exposed. Since otherwise this just stands out like a sore thumb.

**Comment by chrsmith** [comment]
@unknwon PTAL

File: README.md
Comment (bobheadxi) [inline README.md:162]:
```suggestion
much simpler in nature, as most integrations will make use of the Clients API. However, the Accounts API is
```

File: accounts/v1/client.go
Comment (bobheadxi) [inline accounts/v1/client.go:27]: Alright I don't want to keep stacking things on for you, but in #12 we introduced a mechanism to make it easier to configure SAMS on behalf of clients, for example to leverage internal networking by changing `samsHost` without having to double-check how hosts are configured and avoid accidentally breaking something. Could you look into reconciling the two mechanisms, maybe by moving the conn config to a standalone `sams.ConnConfig` type that can be used across all SAMS clients (`clients/v1`, `accounts/v1`?)

Comment (bobheadxi) [reply accounts/v1/client.go:27]: I can make a follow up PR too, if you don't mind refactoring your call sites!

Comment (chrsmith) [reply accounts/v1/client.go:27]: I would not mind! (But I would like to get this PR merged so I can get started on the SSC-integration to use it.) I'm assuming the env var you will load is `SAMS_URL`? That's the one we use for connecting to the Clients API in the SSC backend.

Comment (bobheadxi) [reply accounts/v1/client.go:27]: There's several configurations:

https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/blob/b228ef93f415345ed8690f2bc75eab3c8563233c/clientv1.go#L57-L66

Maybe not all of them apply here, but the big ones are `SAMS_URL` and `SAMS_API_URL`

**Review by bobheadxi — APPROVED** [review]
Approve to unblock, I will follow up on unifying config: https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/13#discussion_r1579738315

**Pull request merged**
